### ISSUE TITLE
Spell checking using browser spell checker

### ISSUE THREAD

**Issue by AndersMad** [microsoft/monaco-editor#1977]
If the DIVs and SPANs is explicitly set to `contenteditable="true"` (instead of inherited) you get a red squiggly under badly spelled words in the monaco editor. The context menu hides the suggestions - but if that is removed/disabled - the word can be replaced; but even without the replace - the spell lint would be very nice.

Such a "feature" would be VERY nice for .TXT, .MD etc... especially with Chromes "new" enhanced (opt. in) spell checker that also "checks the language".

**Comment by dmsnell** [comment]
Hi @AndersMad - do you have a patch for this or know what the implications are for editing, cursor movement, and text replacement? I'm very interested in having native spell-checking enabled while still using Monaco…

**Comment by AndersMad** [comment]
Hi @dmsnell, no - I quickly concluded that this would take some inner workings for vscode and I "only" needed it for some MD editing.. But please do upvote it here - I'd think/guess that it would be fairly easy to achieve for a core developer.

**Comment by dmsnell** [comment]
Thanks for the quick response. I was curious because I assumed the opposite - that it might involve some major overhauls 😄 

It was surprising to me to find that it's not possible. I hope there is something we're all missing.

**Comment by elliott-wen** [comment]
My strategy is to do the spell checking via a language server. I have to admit that it is a bit clumsy. I also have to implement the `add to dictionary` function. It would be great if monaco could leverage built-in spell checkers in browsers.

**Comment by omerdun** [comment]
@elliott-wen I am facing the same issue of trying to implement spell checking to a monaco in a browser. Mind sharing how you ended up solving that?

**Comment by mrinterweb** [comment]
Really wishing spell checking was available. 👍🏻 hope this feature lands soon.

**Comment by Jirido** [comment]
I suport this suggestion, Not having access to a spellcheck when writing the readme.md is a bad experience.

**Comment by abn0mad** [comment]
+1 on adding spell checking using built-in browser spell checking. Being able to use LanguageTool or Grammarly when editing markdown or plain text would really make Monaco a lot more useful.

**Comment by tc1989tc** [comment]
Does it have a plan to implement spell-checking in Monaco? I see codemirror supports the Grammarly web extension

**Comment by batchfly** [comment]
I second the spelling check feature as it would be really helpful for those who use monaco for writing natural languages like documents and comments.

And I agree with @AndersMad  that making Monaco compatible with Grammarly is not that hard for core developer. Editors like codemirror naturally support grammarly.

**Comment by purocean** [comment]
I created a simple library to address this need.

Demo: https://purocean.github.io/monaco-spellchecker/
Code: https://github.com/purocean/monaco-spellchecker/

**Comment by batchor** [comment]
> I created a simple library to address this need.
> 
> Demo: https://purocean.github.io/monaco-spellchecker/ Code: https://github.com/purocean/monaco-spellchecker/

Great work and I'll definitely give it a shot.

Can you explain a little bit (either here or in the project readme) how the new library works?

**Comment by batchor** [comment]
> > > I created a simple library to address this need.
> > > Demo: https://purocean.github.io/monaco-spellchecker/ Code: https://github.com/purocean/monaco-spellchecker/
> > 
> > 
> > Great work and I'll definitely give it a shot.
> > Can you explain a little bit (either here or in the project readme) how the new library works?
> 
> I wanted to introduce spell-checking functionality into my application, [yn](https://github.com/purocean/yn). After researching available solutions, I couldn't find a suitable one, so I decided to implement my own. I spent an afternoon working on this, and the result is a simple version of spell-checking. To achieve this, I used [Typo.js](https://github.com/cfinke/Typo.js) for the spell-checking logic and integrated it with Monaco Editor using the `monaco.editor.setModelMarkers` API for visual decorations and `monaco.languages.registerCodeActionProvider` for providing spelling corrections.
> 
> This is just a basic implementation, and there may be other issues yet to be discovered and addressed.

Thanks for the clarification. Is it possible to allow 3rd-party for spelling checking and correction?

There are a lot spelling checking plugins, e.g. grammarly, and chrome itself has its own spelling checking and correction.

It would be easier and more focused to just exposed the monaco text to 3rd-party and let them do the checking and correction.

The current situation is that monaco seems do not expose its text and 3rd-party tools cannot detect/correct the spelling errors.

**Comment by stevemk14ebr** [comment]
@batchor I don't think so. If you look at the DOM vscode add and removes DOM content as the user scrolls. This alone prevents third party things from inspecting the DOM to find spelling mistakes.
